File: src/foam/dao/ORBitalDAO.java
package foam.dao;

import foam.core.ClassInfo;
import foam.core.FObject;
import foam.core.MethodInfo;
import foam.core.X;
import foam.mlang.order.Comparator;
import foam.mlang.predicate.Predicate;

/**
 * DAO class used to delegate a method execution request to the specified object (specified by an ORBRequest object).
 */
public class ORBitalDAO extends foam.dao.AbstractDAO{

  /**
   * Method that delegates execution to method of object specified by {@param obj}
   *
   * @param obj argument must be an ORBRequest object
   */
  @Override
  public Object cmd_(foam.core.X x, Object obj){
    ORBRequest request = (ORBRequest) obj;
    String receiverID = request.getReceiverObjectID();

    // Find the object if the request is by ID
    FObject receiverObj =  ( receiverID == null )? request.getReceiverObject() : (FObject) x.get(request.getReceiverObjectID());

    ClassInfo classInfo = receiverObj.getClassInfo();
    MethodInfo methodInfo = (MethodInfo) classInfo.getAxiomByName(request.getMethodName());

    return methodInfo.call(x, receiverObj, request.getArgs());
  }

  @Override
  public FObject put_(X x, FObject obj) {
    return null;
  }

  @Override
  public FObject remove_(X x, FObject obj) {
    return null;
  }

  @Override
  public FObject find_(X x, Object id) {
    return null;
  }

  @Override
  public Sink select_(X x, Sink sink, long skip, long limit, Comparator order, Predicate predicate) {
    return null;
  }

}

File: src/foam/dao/ORBitalDAO.js
/**
 * @license
 * Copyright 2019 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

foam.CLASS({
  package: 'foam.dao',
  name: 'ORBitalDAO',
  documentation : 'DAO class used to delegate a method execution request to the specified object (specified by an ORBRequest object).',
  extends: 'foam.dao.ProxyDAO',
  flags: ['java'],

  javaImports: [
    'foam.core.FObject',
    'foam.core.ClassInfo',
    'foam.core.MethodInfo',
    'foam.core.X',
    'foam.nanos.logger.Logger'
  ],

  properties: [

  ],

  methods: [
    {
      name: 'cmd_',
      type: 'Object',

      args: [
        { name: 'x' , type: 'foam.core.X'}, { name: 'obj' , type: 'Object' }
      ],

      javaCode: `

        if ( ! ( obj instanceof ORBRequest ) ) return getDelegate().cmd_(x, obj);
        ORBRequest request = (ORBRequest) obj;

        String receiverID = request.getReceiverObjectID();

        // Get the receiving object
        FObject receiverFObj = request.getReceiverObject();

        receiverFObj =  ( receiverFObj == null && receiverID != null )? (FObject) x.get(receiverID) : receiverFObj;

        if ( receiverFObj == null ) return null;

        ClassInfo classInfo = receiverFObj.getClassInfo();
        MethodInfo methodInfo = (MethodInfo) classInfo.getAxiomByName(request.getMethodName());
        return methodInfo.call(x, receiverFObj, request.getArgs());

      `
    }
  ]

});

